File: CS/ColumnTemplateSelectorExample/ViewModel.cs
using System;
using System.ComponentModel;
using System.Xml.Serialization;
using System.Collections.Generic;
using System.Reflection;
using DevExpress.Xpf.Grid;
using System.ComponentModel.DataAnnotations;

namespace Model {
    public class ViewModel {
        public List<int> Ids { get; private set; }
        public List<string> Cities { get; private set; }
        public List<string> Titles { get; private set; }
        public IList<Employee> Source { get; private set; }
        public ViewModel() {
            Source = EmployeesData.DataSource;
            List<string> cities = new List<string>();
            foreach(Employee employee in Source) {
                if(!cities.Contains(employee.City))
                    cities.Add(employee.City);
            }
            Cities = cities;
            Titles = new List<string>() { "Mr.", "Ms." };
            List<int> ids = new List<int>();
            foreach(Employee employee in Source) {
                if(!ids.Contains(employee.Id))
                    ids.Add(employee.Id);
            }
            Ids = ids;
        }
    }

    [XmlRoot("Employees")]
    public class EmployeesData : List<Employee> {
        public static IList<Employee> DataSource {
            get {

                XmlSerializer s = new XmlSerializer(typeof(EmployeesData));
#if SILVERLIGHT
                return (List<Employee>)s.Deserialize(Assembly.GetExecutingAssembly().GetManifestResourceStream("SLColumnTemplateSelectorExample.EmployeesWithPhoto.xml"));
#else
                return (List<Employee>)s.Deserialize(Assembly.GetExecutingAssembly().GetManifestResourceStream("ColumnTemplateSelectorExample.EmployeesWithPhoto.xml"));
#endif
            }
        }
    }

    public class Employee {
        [ColumnGeneratorTemplateNameAttribute("IdTemplate")]
        public int Id { get; set; }
        public string FirstName { get; set; }
        [ColumnGeneratorTemplateNameAttribute("CityTemplate")]
        public string City { get; set; }
        public DateTime BirthDate { get; set; }
        [ColumnGeneratorTemplateNameAttribute("TitleTemplate")]
        public string Title { get; set; }
        [ColumnGeneratorTemplateNameAttribute("ImageTemplate"), Display(Name = "Photo")]
        public byte[] ImageData { get; set; }
    }
}

File: VB/ColumnTemplateSelectorExample/ViewModel.vb
Imports System.ComponentModel
Imports System.Xml.Serialization
Imports System.Collections.Generic
Imports System.Reflection
Imports DevExpress.Xpf.Grid
Imports System.ComponentModel.DataAnnotations

Namespace Model

    Public Class ViewModel

        Private _Ids As List(Of Integer), _Cities As List(Of String), _Titles As List(Of String), _Source As IList(Of Model.Employee)

        Public Property Ids As List(Of Integer)
            Get
                Return _Ids
            End Get

            Private Set(ByVal value As List(Of Integer))
                _Ids = value
            End Set
        End Property

        Public Property Cities As List(Of String)
            Get
                Return _Cities
            End Get

            Private Set(ByVal value As List(Of String))
                _Cities = value
            End Set
        End Property

        Public Property Titles As List(Of String)
            Get
                Return _Titles
            End Get

            Private Set(ByVal value As List(Of String))
                _Titles = value
            End Set
        End Property

        Public Property Source As IList(Of Employee)
            Get
                Return _Source
            End Get

            Private Set(ByVal value As IList(Of Employee))
                _Source = value
            End Set
        End Property

        Public Sub New()
            Source = EmployeesData.DataSource
            Dim cities As List(Of String) = New List(Of String)()
            For Each employee As Employee In Source
                If Not cities.Contains(employee.City) Then cities.Add(employee.City)
            Next

            Me.Cities = cities
            Titles = New List(Of String)() From {"Mr.", "Ms."}
            Dim ids As List(Of Integer) = New List(Of Integer)()
            For Each employee As Employee In Source
                If Not ids.Contains(employee.Id) Then ids.Add(employee.Id)
            Next

            Me.Ids = ids
        End Sub
    End Class

    <XmlRoot("Employees")>
    Public Class EmployeesData
        Inherits List(Of Employee)

        Public Shared ReadOnly Property DataSource As IList(Of Employee)
            Get
                Dim s As XmlSerializer = New XmlSerializer(GetType(EmployeesData))
#If SILVERLIGHT
                return (List<Employee>)s.Deserialize(Assembly.GetExecutingAssembly().GetManifestResourceStream("SLColumnTemplateSelectorExample.EmployeesWithPhoto.xml"));
#Else
                Return CType(s.Deserialize(Assembly.GetExecutingAssembly().GetManifestResourceStream("ColumnTemplateSelectorExample.EmployeesWithPhoto.xml")), List(Of Employee))
#End If
            End Get
        End Property
    End Class

    Public Class Employee

        <ColumnGeneratorTemplateNameAttribute("IdTemplate")>
        Public Property Id As Integer

        Public Property FirstName As String

        <ColumnGeneratorTemplateNameAttribute("CityTemplate")>
        Public Property City As String

        Public Property BirthDate As Date

        <ColumnGeneratorTemplateNameAttribute("TitleTemplate")>
        Public Property Title As String

        <ColumnGeneratorTemplateNameAttribute("ImageTemplate"), Display(Name:="Photo")>
        Public Property ImageData As Byte()
    End Class
End Namespace
